File: apps/clients/python/agent.py
import logging
import threading
from typing import Optional

from config import Config
from device import DevicePayload
from mqtt_client import MQTTClient, ActionEnvelope, connect_mqtt
from subsystems import SubsystemSet
from vento_client import VentoClient

LOG = logging.getLogger(__name__)


class Agent:
    def __init__(
        self,
        cfg: Config,
        client: VentoClient,
        skip_register_actions: bool = False,
        run_once: bool = False,
    ) -> None:
        self.cfg = cfg
        self.http = client
        self.skip_register_actions = skip_register_actions
        self.run_once = run_once
        self.subs = SubsystemSet(cfg)
        self.stop_event = threading.Event()
        self.mqtt: Optional[MQTTClient] = None

    def start(self) -> None:
        self.subs.prepare(self.cfg.device_name)
        self._ensure_device()
        self.mqtt = connect_mqtt(
            hostname=self.http.hostname(),
            device_name=self.cfg.device_name,
            username=self.cfg.username,
            token=self.cfg.token,
            handler=self._handle_action,
        )
        LOG.info("connected to mqtt as %s", self.cfg.device_name)
        try:
            self.subs.publish_boot(self.mqtt)
            if self.run_once:
                return
            self.subs.start_intervals(self.mqtt, self.stop_event)
            self.stop_event.wait()
        finally:
            if self.mqtt:
                self.mqtt.close()

    def stop(self) -> None:
        self.stop_event.set()

    def _ensure_device(self) -> None:
        payload: DevicePayload = self.subs.device_payload(self.cfg.device_name)
        exists = self.http.device_exists(self.cfg.token, self.cfg.device_name)
        if exists:
            self.http.update_device(self.cfg.token, self.cfg.device_name, payload.to_dict())
            return

        LOG.info("device %s not found, registering...", self.cfg.device_name)
        create_payload = {"name": payload.name, "platform": payload.current_sdk}
        self.http.register_device(self.cfg.token, create_payload)
        self.http.update_device(self.cfg.token, self.cfg.device_name, payload.to_dict())
        if not self.skip_register_actions:
            try:
                self.http.trigger_register_actions(self.cfg.token)
            except Exception as err:
                LOG.warning("failed to trigger registerActions: %s", err)

    def _handle_action(self, msg: ActionEnvelope) -> None:
        handled = self.subs.handle_action(msg)
        if not handled:
            LOG.info("unhandled action[%s/%s]: %s", msg.subsystem, msg.action, msg.payload)

File: apps/clients/go/internal/agent/agent.go
package agent

import (
	"context"
	"fmt"
	"log"

	"ventoagent/internal/config"
	"ventoagent/internal/subsystems"
	"ventoagent/internal/vento"
)

// ConfigWriter abstracts the config persistence used by the agent.
type ConfigWriter interface {
	Save(*config.Config) error
}

// Options configure the Agent.
type Options struct {
	Config              *config.Config
	Client              *vento.Client
	ConfigWriter        ConfigWriter
	SkipRegisterActions bool
	RunOnce             bool
}

// Agent runs the local monitoring loop.
type Agent struct {
	cfg   *config.Config
	http  *vento.Client
	store ConfigWriter

	mqtt *vento.MQTTClient

	subs *subsystems.Set

	skipRegisterActions bool
	runOnce             bool
}

// New builds a new Agent from the provided options.
func New(opts Options) *Agent {
	return &Agent{
		cfg:                 opts.Config,
		http:                opts.Client,
		store:               opts.ConfigWriter,
		subs:                subsystems.NewSet(opts.Config),
		skipRegisterActions: opts.SkipRegisterActions,
		runOnce:             opts.RunOnce,
	}
}

// Start ensures the device exists, connects to MQTT and launches monitors.
func (a *Agent) Start(ctx context.Context) error {
	a.subs.Prepare(a.cfg.DeviceName)

	if err := a.ensureDevice(ctx); err != nil {
		return fmt.Errorf("ensure device: %w", err)
	}

	mqttClient, err := vento.ConnectMQTT(ctx, a.http.BaseURLObject(), a.cfg.DeviceName, a.cfg.Username, a.cfg.Token, a.handleAction)
	if err != nil {
		return fmt.Errorf("mqtt connection failed: %w", err)
	}
	a.mqtt = mqttClient
	defer a.mqtt.Close()
	log.Printf("connected to mqtt broker as %s", a.cfg.DeviceName)

	// Set up reconnection handler to re-publish boot data
	a.mqtt.SetOnReconnect(func() {
		log.Printf("reconnected - re-publishing boot data...")
		a.subs.PublishBoot(ctx, a.mqtt)
	})

	a.subs.PublishBoot(ctx, a.mqtt)
	if a.runOnce {
		return nil
	}

	a.subs.StartIntervals(ctx, a.mqtt)

	<-ctx.Done()
	return ctx.Err()
}

func (a *Agent) ensureDevice(ctx context.Context) error {
	exists, err := a.http.DeviceExists(ctx, a.cfg.Token, a.cfg.DeviceName)
	if err != nil {
		return err
	}
	payload := a.subs.DevicePayload(a.cfg.DeviceName)
	if exists {
		return a.http.UpdateDevice(ctx, a.cfg.Token, a.cfg.DeviceName, payload)
	}
	log.Printf("device %s not found, registering...", a.cfg.DeviceName)
	createPayload := map[string]any{
		"name":     payload.Name,
		"platform": payload.Platform,
	}
	if err := a.http.RegisterDevice(ctx, a.cfg.Token, createPayload); err != nil {
		return err
	}
	if err := a.http.UpdateDevice(ctx, a.cfg.Token, a.cfg.DeviceName, payload); err != nil {
		return err
	}
	if !a.skipRegisterActions {
		if err := a.http.RegenerateBoardForDevice(ctx, a.cfg.Token, a.cfg.DeviceName); err != nil {
			log.Printf("warning: failed to regenerate board: %v", err)
		}
	}
	// persist generated config to disk
	if a.store != nil {
		if err := a.store.Save(a.cfg); err != nil {
			log.Printf("warning: failed to persist config: %v", err)
		}
	}
	return nil
}

func (a *Agent) handleAction(msg vento.ActionEnvelope) {
	if handled := a.subs.HandleAction(msg); handled {
		return
	}
	log.Printf("unhandled action[%s/%s]: %s", msg.Subsystem, msg.Action, string(msg.Payload))
}
